### ISSUE TITLE
"Cannot find module" errors when running latest versions

### ISSUE THREAD

**Issue by luanmm** [pulumi/pulumi#11973]
### What happened?

I'm trying Pulumi in a project for a few months ago, but now I'm receiving an error that I couldn't identify the cause (only a workaround).

Running as always (didn't noticed when exactly it started to happen, if there is something with a version update, code change, etc), using `pulumi up`, it started to give me an "Cannot find module" error.

### Steps to reproduce

Haven't found how to reproduce it in a "clean install" yet.

### Expected Behavior

Build and run the deploy like it always did, without requiring additional dependencies (that is not already present in Pulumi's packages).

### Actual Behavior

First, it started to throw this error:

```bash
Diagnostics:
  pulumi:pulumi:Stack (...):
    error: Running program '/home/.../deploy' failed with an unhandled exception:
    Error: Cannot find module 'json-parse-even-better-errors'
    Require stack:
    - /home/.../deploy/node_modules/read-package-json/read-json.js
    - /home/.../deploy/node_modules/read-package-tree/rpt.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/runtime/index.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/index.js
    - /home/.../deploy/node_modules/@pulumi/awsx/index.js
    - /home/.../deploy/api.ts
    - /home/.../deploy/index.ts
    - /home/.../deploy/node_modules/@pulumi/pulumi/cmd/run/run.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/cmd/run/index.js
        at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
        at Function.Module._load (node:internal/modules/cjs/loader:778:27)
        at Module.require (node:internal/modules/cjs/loader:1005:19)
        at require (node:internal/modules/cjs/helpers:102:18)
        at Object.<anonymous> (/home/.../deploy/node_modules/read-package-json/read-json.js:7:16)
        at Module._compile (node:internal/modules/cjs/loader:1103:14)
        at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        at Module.load (node:internal/modules/cjs/loader:981:32)
        at Function.Module._load (node:internal/modules/cjs/loader:822:12)
        at Module.require (node:internal/modules/cjs/loader:1005:19)
```

But it is very strange, as my project itself never call this library.

So, even not being a need for my project, I referenced this library in the `package.json` (as a "workaround").

Then it started complaining about another dependency:

```bash
Diagnostics:
  pulumi:pulumi:Stack (...):
    error: Running program '/home/.../deploy' failed with an unhandled exception:
    Error: Cannot find module 'npm-normalize-package-bin'
    Require stack:
    - /home/.../deploy/node_modules/read-package-json/read-json.js
    - /home/.../deploy/node_modules/read-package-tree/rpt.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/runtime/index.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/index.js
    - /home/.../deploy/node_modules/@pulumi/awsx/index.js
    - /home/.../deploy/api.ts
    - /home/.../deploy/index.ts
    - /home/.../deploy/node_modules/@pulumi/pulumi/cmd/run/run.js
    - /home/.../deploy/node_modules/@pulumi/pulumi/cmd/run/index.js
        at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
        at Function.Module._load (node:internal/modules/cjs/loader:778:27)
        at Module.require (node:internal/modules/cjs/loader:1005:19)
        at require (node:internal/modules/cjs/helpers:102:18)
        at Object.<anonymous> (/home/.../deploy/node_modules/read-package-json/read-json.js:9:27)
        at Module._compile (node:internal/modules/cjs/loader:1103:14)
        at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        at Module.load (node:internal/modules/cjs/loader:981:32)
        at Function.Module._load (node:internal/modules/cjs/loader:822:12)
        at Module.require (node:internal/modules/cjs/loader:1005:19)
```

After puting these two, it worked like before, as it should.

### Output of `pulumi about`

```bash
CLI          
Version      3.52.1
Go Version   go1.19.5
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.25.0
awsx        1.0.1
cloudflare  4.14.0
docker      3.6.1
kubernetes  3.23.1
nodejs      unknown
random      4.8.2

Host     
OS       debian
Version  bookworm/sid
Arch     x86_64

This project is written in nodejs: executable='/usr/bin/node' version='v16.14.2'

Current Stack: ...

TYPE                                       URN
pulumi:pulumi:Stack                        urn:pulumi:...::...::pulumi:pulumi:Stack::...-...
pulumi:providers:pulumi                    urn:pulumi:...::...::pulumi:providers:pulumi::default
pulumi:providers:awsx                      urn:pulumi:...::...::pulumi:providers:awsx::default_1_0_1
pulumi:providers:cloudflare                urn:pulumi:...::...::pulumi:providers:cloudflare::...
pulumi:providers:random                    urn:pulumi:...::...::pulumi:providers:random::default_4_8_2
random:index/randomString:RandomString     urn:pulumi:...::...::random:index/randomString:RandomString::...
awsx:ecr:Repository                        urn:pulumi:...::...::awsx:ecr:Repository::...
pulumi:providers:aws                       urn:pulumi:...::...::pulumi:providers:aws::default_5_16_2
pulumi:pulumi:StackReference               urn:pulumi:...::...::pulumi:pulumi:StackReference::.../.../...
pulumi:providers:kubernetes                urn:pulumi:...::...::pulumi:providers:kubernetes::provider
kubernetes:core/v1:Secret                  urn:pulumi:...::...::kubernetes:core/v1:Secret::...
kubernetes:core/v1:Service                 urn:pulumi:...::...::kubernetes:core/v1:Service::...
kubernetes:core/v1:Service                 urn:pulumi:...::...::kubernetes:core/v1:Service::...
kubernetes:cert-manager.io/v1:Certificate  urn:pulumi:...::...::kubernetes:cert-manager.io/v1:Certificate::...
kubernetes:core/v1:Secret                  urn:pulumi:...::...::kubernetes:core/v1:Secret::...
kubernetes:apps/v1:StatefulSet             urn:pulumi:...::...::kubernetes:apps/v1:StatefulSet::...
aws:ecr/repository:Repository              urn:pulumi:...::...::awsx:ecr:Repository$aws:ecr/repository:Repository::...
aws:ecr/lifecyclePolicy:LifecyclePolicy    urn:pulumi:...::...::awsx:ecr:Repository$aws:ecr/lifecyclePolicy:LifecyclePolicy::...
awsx:ecr:Image                             urn:pulumi:...::...::awsx:ecr:Image::...
kubernetes:apps/v1:Deployment              urn:pulumi:...::...::kubernetes:apps/v1:Deployment::...
kubernetes:apps/v1:Deployment              urn:pulumi:...::...::kubernetes:apps/v1:Deployment::...
kubernetes:core/v1:Service                 urn:pulumi:...::...::kubernetes:core/v1:Service::...
kubernetes:networking.k8s.io/v1:Ingress    urn:pulumi:...::...::kubernetes:networking.k8s.io/v1:Ingress::...
kubernetes:networking.k8s.io/v1:Ingress    urn:pulumi:...::...::kubernetes:networking.k8s.io/v1:Ingress::...
cloudflare:index/record:Record             urn:pulumi:...::...::cloudflare:index/record:Record::...
cloudflare:index/record:Record             urn:pulumi:...::...::cloudflare:index/record:Record::...


Found no pending operations associated with ...

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/...
User           ...
Organizations  ..., ...

Dependencies:
NAME                              VERSION
@pulumi/aws                       5.25.0
@pulumi/awsx                      1.0.1
@pulumi/cloudflare                4.14.0
@pulumi/kubernetes                3.23.1
@pulumi/kubernetesx               0.1.6
@pulumi/pulumi                    3.50.2
@pulumi/random                    4.8.2
uuid                              9.0.0
@types/node                       16.18.11
@types/uuid                       9.0.0
@typescript-eslint/eslint-plugin  5.47.1
@typescript-eslint/parser         5.47.1
eslint                            8.30.0
eslint-config-prettier            8.5.0
eslint-plugin-prettier            4.2.1
json-parse-even-better-errors     3.0.0
npm-normalize-package-bin         3.0.0
prettier                          2.8.1
typescript                        4.9.4

Pulumi locates its logs in /tmp by default
```

### Additional context

This is the `package.json` used:

```json
{
  "name": "...",
  "main": "index.ts",
  "devDependencies": {
    "@types/node": "^16",
    "@types/uuid": "^9.0.0",
    "@typescript-eslint/eslint-plugin": "^5.47.1",
    "@typescript-eslint/parser": "^5.47.1",
    "eslint": "^8.30.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-prettier": "^4.2.1",
    "json-parse-even-better-errors": "^3.0.0", // These two are needed here to make it work
    "npm-normalize-package-bin": "^3.0.0", // (but shouldn't, as neither are dependencies of my project)
    "prettier": "^2.8.1",
    "typescript": "~4.9.4"
  },
  "dependencies": {
    "@pulumi/aws": "^5.25.0",
    "@pulumi/awsx": "^1.0.1",
    "@pulumi/cloudflare": "^4.14.0",
    "@pulumi/kubernetes": "^3.23.1",
    "@pulumi/kubernetesx": "^0.1.6",
    "@pulumi/pulumi": "^3.50.2",
    "@pulumi/random": "^4.8.2",
    "uuid": "^9.0.0"
  }
}
```

Another detail is that I'm using Yarn 1.22.19 as package manager.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
Extra odd because that seems to be coming from the `read-package-tree` dependency which we haven't changed in years, and is definitely included in the `@pulumi/pulumi` package dependencies.

> Haven't found how to reproduce it in a "clean install" yet.

Does this mean if you clear and re-install node_modules without the two additional dependencies it works correctly?

If so I'm inclined to think you somehow corrupted that node_modules state between uses.

**Comment by luanmm** [comment]
Hello, @Frassle. Thank you for the reply.

Sorry if I couldn't express myself clearly, but what I meant with "clean install" is because I just got this error in a project that is being incrementally changing over the time and this error started without I even noticing what could be causing it. Is not occurring when I try to emulate it isolating from my project (but I did not spend a lot of time on this too).

Your thoughts pointed me to a new try here: clean everything (including `node_modules` and `yarn.lock`) and try to install again. So Yarn pointed out what may be a cause for it (I think):

```bash
$ yarn install
yarn install v1.22.19
warning package.json: No license field
info No lockfile found.
warning ...: No license field
[1/4] Resolving packages...
warning @pulumi/aws > read-package-tree@5.3.1: The functionality that this package provided is now in @npmcli/arborist
warning @pulumi/aws > read-package-tree > readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
warning @pulumi/aws > aws-sdk > querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
warning @pulumi/aws > aws-sdk > url > querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
warning @pulumi/aws > @pulumi/pulumi > read-package-tree@5.3.1: The functionality that this package provided is now in @npmcli/arborist
warning @pulumi/aws > @pulumi/pulumi > @opentelemetry/instrumentation-grpc > @opentelemetry/api-metrics@0.32.0: Please use @opentelemetry/api >= 1.3.0
warning @pulumi/aws > @pulumi/pulumi > @opentelemetry/instrumentation-grpc > @opentelemetry/instrumentation > @opentelemetry/api-metrics@0.32.0: Please use @opentelemetry/api >= 1.3.0
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 82.89s.
```

If the dependencies that are needed here in my `package.json` come from `read-package-tree`, like you stated, this warnings that appeared can give us a clue why those errors are occurring.

**Comment by Frassle** [comment]
Does the project work after clearing node_modules and re-running yarn install?

Those warnings are just warnings, we see them in our tests as well, it should still be installing the read-package-tree modules and it's dependencies.

**Comment by luanmm** [comment]
I tried to remove the dependencies and it worked now without them. I think my lock-file got corrupted at some point, indeed (maybe when updating the dependecies). If so, the error might be related to Yarn itself. Thanks for pointing out to the solution.

Anyway, besides the warnings not impacting directly in errors like these one, it could be addressed to avoid future issues. I've seen a lot in Pulumi-related libraries (there are some in "beta", with versions below 1.0, so it is acceptable yet).

Congratulations for the work done with Pulumi. I'm just a tester for now, but I've liked a lot the platform's design principles and the way it works in practice. Hope this project could follow a successful path!

I will close the issue, as it seems not to be related to Pulumi directly. Thank you for your time (and sorry that I've pointed out to the wrong direction with this issue here).

Best regards.
